### PR TITLE
add new feature for llava-next-video

### PR DESCRIPTION
**Pull request by ZhangYuanhan-AI**
llava-next-video: https://llava-vl.github.io/blog/2024-04-30-llava-next-video/

### CODE REVIEW

File: python/sglang/utils.py [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
 import base64
 import json
+import os
 import threading
 import urllib.request
+from concurrent.futures import ThreadPoolExecutor
 from io import BytesIO
 from json import dumps
 
+import cv2
```
Comment (hkunzhe) [inline python/sglang/utils.py:12]: `opencv-python` should be added in `pyproject.toml`.

File: examples/quick_start/srt_example_llava_v.py
Comment (hkunzhe) [inline examples/quick_start/srt_example_llava_v.py:1]: There are many user-defined paths and some parameters in argparse are not being used.

File: README.md [hunk 2]
```diff
@@ -1,411 +1,74 @@
+To refine and clean up the README you've provided for the SGLang project, I'll focus on improving clarity, organization, and conciseness. This includes providing clear installation instructions, simplifying steps where possible, and ensuring the document is easy to navigate. Here's a revised version:
```
Comment (merrymercy) [inline README.md]: Revert the change to README.md

File: python/pyproject.toml [hunk 3]
```diff
@@ -20,7 +20,7 @@ dependencies = [
 
 [project.optional-dependencies]
 srt = ["aiohttp", "fastapi", "psutil", "rpyc", "torch", "uvloop", "uvicorn",
-       "zmq", "vllm>=0.3.3", "interegular", "pydantic", "pillow", "outlines>=0.0.27"]
+       "zmq", "vllm==0.3.3", "interegular", "pydantic", "pillow", "outlines>=0.0.27"]
```
Comment (merrymercy) [inline python/pyproject.toml]: Recently, we moved to vllm==0.4.2, can you update accordingly?
There are only minor changes. You can change similar to the changes in this PR https://github.com/sgl-project/sglang/pull/380

File: examples/demo/model_call.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
+# from flask import Flask, request, jsonify
```
Comment (merrymercy) [inline examples/demo/model_call.py]: Rename `examples/demo` to `examples/usage/llava_video`

File: examples/quick_start/srt_example_llava_v.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        average_time = total_time / len(batch_video_files)
+        print(f"Number of videos in batch: {len(batch_video_files)}. Average processing time per video: {average_time:.2f} seconds. Total time for this batch: {total_time:.2f} seconds")
+
+        save_batch_results(batch_video_files, states, cur_chunk, num_batches, save_dir)
+        num_batches += 1
+
+    compile_and_cleanup_final_results(cur_chunk, num_batches, save_dir)
+
+
+import argparse
```
Comment (merrymercy) [inline examples/quick_start/srt_example_llava_v.py]: move this to the beginning of the file.

File: examples/quick_start/srt_example_llava_v.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+    # Create the parser
+    parser = argparse.ArgumentParser(description='Run video processing with specified port.')
+
+    # Add an argument for the port
+    parser.add_argument('--port', type=int, default=30000, help='The master port for distributed serving.')
+    parser.add_argument('--chunk-idx', type=int, default=0, help='The index of the chunk to process.')
+    parser.add_argument('--num-chunks', type=int, default=8, help='The number of chunks to process.')
+    parser.add_argument('--save-dir', type=str, default="./work_dirs/llava_video", help='The directory to save the processed video files.')
+    parser.add_argument('--video-dir', type=str, default="/mnt/bn/vl-research/workspace/yhzhang/data/sora/", help='The directory to save the processed video files.')
```
Comment (merrymercy) [inline examples/quick_start/srt_example_llava_v.py]: It seems there are many personal paths. Can someone else easily run this script?
Maybe you can move all things under a single folder `examples/llava_video` and add a README.md with an example command that runs on your Q98Z4OTh8RwmDonc.mp4

File: python/sglang/utils.py [hunk 7]
```diff
@@ -130,6 +137,107 @@ def encode_image_base64(image_path):
...
+    # Concatenate all frames bytes
+    video_bytes = b"".join(encoded_frames)
+
+    # Encode the concatenated bytes to base64
+    video_base64 = "video:" + base64.b64encode(video_bytes).decode("utf-8")
+
+    return video_base64
+
+
+# def encode_video_base64(video_path):
```
Comment (merrymercy) [inline python/sglang/utils.py]: remove unused code.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/router/model_rpc.py [hunk 8]
```diff
@@ -32,6 +32,9 @@
 )
 from vllm.logger import _default_handler as vllm_default_handler
 
+from typing import Any, Dict, List, Optional, Tuple, Union
```
Comment (merrymercy) [inline python/sglang/srt/managers/router/model_rpc.py]: merge this with L7

**Pull request closed without merging**